src/passes/OptimizeInstructions.cpp

```diff
@@ -1272,16 +1284,26 @@ struct OptimizeInstructions
       return right;
     }
     // x == 0   ==>   eqz x
-    if ((matches(curr, binary(Abstract::Eq, any(&left), ival(0))))) {
-      return builder.makeUnary(EqZInt64, left);
+    if (matches(curr, binary(Abstract::Eq, any(&left), ival(0)))) {
+      return builder.makeUnary(Abstract::getUnary(left->type, Abstract::EqZ),
+                               left);
     }
-
     // Operations on one
     // (signed)x % 1   ==>   0
     if (matches(curr, binary(Abstract::RemS, pure(&left), ival(1)))) {
       right->value = Literal::makeSingleZero(type);
       return right;
     }
+    // (signed)x % C_pot != 0   ==>  x & (C_pot - 1) != 0
+    if (matches(curr,
+                binary(Abstract::Ne,
+                       binary(&inner, Abstract::RemS, any(&left), constant(&c)),
```

Comment:
```suggestion
                       binary(&inner, Abstract::RemS, any(&left), ival()),
```
You actually don't need `c` here because it is already captured in `right`.

Reply:
right!

Reply:
Hmm, I got regression

Reply:
Yes, `right` and `c` are different constants. `right` is outer constant and `c` is inner for this expression:
```
(signed)x % c != 0   ==>  x & (c - 1) != 0
```
where `0` is `right`

Reply:
Oh right, my bad!